Added the Best Practices section with general principles. (#5137)
istio/istio.io: content/en/docs/best-practices/general-principles/index.md

---

> title: General Principles

Reviewer: This seems more like "Multicluster Principles" than general to me?

Reviewer: I would think a general principles one would be more along the lines of make sure you have proper HA setup (with info about how to do this), fine tuning resource requests/limits, etc

Reviewer 2: Agreed, these are really deployment principles rather than general best practice principles.

I think a few folks have pointed this out.  @rcaballeromx should we rename this doc?

Author: @nmittler @howardjohn would this suggestion work better?
```suggestion
title: Deployment Best Practices
```

description: The general principles for your Istio deployments.
weight: 1
icon: best-practices
---

We've identified the following general principles to help you get the most out
of your Istio deployments.

> ## Make your deployment simple

Reviewer 3: Can we just call this "Prefer simplicity" or something?

Although reading it, I'm not actually sure what it really means. What does it mean to prefer simplicity in deployment model? Can we be much more explicit here, and say that we should prefer using as few clusters as possible, as few networks as possible, and as few meshes as possible?

Reviewer 4: @smawson when I think about multicluster, I think we have 6 dimensions - that can range from 1..n.  Clearly 6 dimensions to infinity is a really big number, and your proposal of using as few of each dimension lines up with our ideas around compressing dimensions into something that is ultimately manageable.  I'm not sure i have helped this conversation along, but for @rcaballeromx this information may be helpful.

Cheers
-steve

Reviewer 2: @smawson 
I think that the attempt here is what we had described previously as "prefer the simplest deployment possible" or "deploy the fewest number of `things`". I think it's a good principle in general, but could probably be reworded to be more clear.

Author: I'm updating the heading to try to capture the feedback. I landed on:
```suggestion
## Prefer simplicity
```

> When choosing a [deployment model](/docs/concepts/deployment-models),

Author: I'm also changing the text to reflect the context you shared, @sdake

**prefer simplicity over complexity** wherever possible.

> ## Make your deployment fast

Reviewer 3: This feels weird as a statement. It isn't really making your deployment fast, which sounds like something about getting a good CI/CD pipeline in place.

This is about favoring geographic proximity to users, maybe we can say something about deploying close to your users?

BTW I get the nice "Make your deployment X" but I don't think it is works as well as having each section be more explicit:
- Prefer simpler deployment models
- Place clusters near your users
- Use multiple availability zones

Reviewer 5: +1 to what Sven said. Saying "Make your deployment fast" as a best practice is a bit redundant. It's not like operator are going to say "Gosh, I wish I would have thought of that, I was trying to make my deployment slow" :-)

Instead, provide a list of specific actions and for each explain the benefits.

Include clusters in your deployment across the globe for **geographic
proximity to end-users**. The proximity helps your deployment have low latency.

## Make your deployment reliable

Include clusters in your deployment **across multiple availability regions
and zones** within each region. This approach helps you avoid global failures.

> Follow these recommendations to enhance the reliability of your deployment:

Reviewer 3: Do we want to say anything about networks? Deploy the control plane to at least one cluster per network is definitely best practice.

Reviewer 5: Isn't at least one control plane per network a requirement? That is, more than a best practice?

Reviewer 2: @rcaballeromx I don't think this bulleted list belongs in principles, which is the source of this confusion. You're effectively re-stating the actual best practices which will come next.  I think if you just remove this for now you're fine.

- Deploy each workload to at least one cluster per region.
- Deploy a control plane to at least one cluster per region.

> - Ensure each control plane has a dedicated configuration source.

Reviewer: This is a bit vague, I am not sure I understand what this means.

In general I think the whole page is a bit high level without explaining how and why to do the things we recommend

Reviewer 2: @rcaballeromx again ... remove this list.

> - Roll out changes incrementally. Do not change all failure domains at once.

Reviewer 3: Change one failure domain at a time? A subset of failure domains? Not sure what best practice actually is, what does the SRE book say?  :)

Reviewer 5: ```suggestion
- Roll out Istio and Kubernetes configuration changes incrementally, one failure domain at a time.
```

Reviewer: Can we make something a best practice if we don't actually support it today? There is not really any (good) way to roll things out incrementally with Istio today

Reviewer 2: @howardjohn that's a good point. What are folks doing out in the wild for this? They must make this work somehow, no?

Reviewer: Fingers crossed things work :slightly_smiling_face: 

Generally people are doing canaries by cluster, and upgrading on a dev cluster

> - Roll out changes incrementally. Do not change all failure domains at once.

Reviewer 3: Also for incrementally, we should make sure we explicitly say that this is for both configuration and binary changes, and both application and Istio changes. Basically everything should be incremental.

> For example, an Istio deployment that satisfies the reliability recommendations above

Reviewer 2: Let's make the example one control plane per cluster. This is likely going to be are default best practice for control planes as it limits the blast radius even further to a single cluster. In addition CI/CD systems will likely operate on a per-cluster basis, which will work well here.

running in Kubernetes would meet the following conditions:

> - The deployment runs a control plane in one cluster per availability zone.

Reviewer 5: Is availability zone == failure domain? 

Do we need to add terms to the glossary?

Reviewer 2: +1
